File: Lab4/apz-pzpi-22-9-aleksiienko-yevhenii-lab4/models/SensorData.js
const mongoose = require('mongoose');

const sensorDataSchema = new mongoose.Schema({
  deviceId: {
    type: String,
    required: true,
    ref: 'Device'
  },
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  sensorType: {
    type: String,
    enum: ['temperature', 'motion', 'humidity', 'system'],
    required: true
  },
  data: {
    // Temperature sensor data
    temperature: {
      type: Number,
      min: -50,
      max: 100
    },
    humidity: {
      type: Number,
      min: 0,
      max: 100
    },
    
    // Motion sensor data
    motionDetected: {
      type: Boolean
    },
    sensorId: {
      type: Number
    },
    location: {
      type: String,
      trim: true
    },
    
    // System data
    battery: {
      type: Number,
      min: 0,
      max: 100
    },
    signalStrength: {
      type: Number,
      min: -100,
      max: 0
    },
    memory: {
      used: Number,
      free: Number
    },
    uptime: {
      type: Number
    }
  },
  alertLevel: {
    type: String,
    enum: ['info', 'warning', 'critical', 'normal'],
    default: 'normal'
  },
  processed: {
    type: Boolean,
    default: false
  },
  processingResult: {
    anomaly: {
      type: Boolean,
      default: false
    },
    confidence: {
      type: Number,
      min: 0,
      max: 1
    },
    recommendations: [{
      type: String,
      maxlength: 500
    }],
    trend: {
      type: String,
      enum: ['increasing', 'decreasing', 'stable', 'volatile']
    }
  },
  metadata: {
    firmware: {
      type: String
    },
    deviceTimestamp: {
      type: Number
    },
    processingTime: {
      type: Number // milliseconds
    },
    dataQuality: {
      type: String,
      enum: ['excellent', 'good', 'poor', 'invalid'],
      default: 'good'
    }
  }
}, {
  timestamps: true
});

// Indexes for performance
sensorDataSchema.index({ deviceId: 1, createdAt: -1 });
sensorDataSchema.index({ userId: 1, createdAt: -1 });
sensorDataSchema.index({ sensorType: 1, createdAt: -1 });
sensorDataSchema.index({ alertLevel: 1, createdAt: -1 });
sensorDataSchema.index({ processed: 1 });
sensorDataSchema.index({ 
  deviceId: 1, 
  sensorType: 1, 
  createdAt: -1 
});

// TTL index for automatic data cleanup (optional - remove old data after 1 year)
sensorDataSchema.index({ createdAt: 1 }, { expireAfterSeconds: 365 * 24 * 60 * 60 });

// Virtual for getting human-readable timestamp
sensorDataSchema.virtual('readableTimestamp').get(function() {
  return this.createdAt.toLocaleString();
});

// Method to mark as processed
sensorDataSchema.methods.markProcessed = function(result = {}) {
  this.processed = true;
  this.processingResult = { ...this.processingResult, ...result };
  return this.save();
};

// Static method to get latest data by device and sensor type
sensorDataSchema.statics.getLatest = function(deviceId, sensorType, limit = 1) {
  return this.find({ deviceId, sensorType })
    .sort({ createdAt: -1 })
    .limit(limit);
};

// Static method to get data in time range
sensorDataSchema.statics.getInTimeRange = function(deviceId, startDate, endDate, sensorType = null) {
  const query = {
    deviceId,
    createdAt: {
      $gte: startDate,
      $lte: endDate
    }
  };
  
  if (sensorType) {
    query.sensorType = sensorType;
  }
  
  return this.find(query).sort({ createdAt: 1 });
};

// Static method to get unprocessed data
sensorDataSchema.statics.getUnprocessed = function(limit = 100) {
  return this.find({ processed: false })
    .sort({ createdAt: 1 })
    .limit(limit);
};

// Static method to get alerts
sensorDataSchema.statics.getAlerts = function(deviceId = null, alertLevel = null, limit = 50) {
  const query = {};
  
  if (deviceId) query.deviceId = deviceId;
  if (alertLevel) query.alertLevel = alertLevel;
  else query.alertLevel = { $in: ['warning', 'critical'] };
  
  return this.find(query)
    .sort({ createdAt: -1 })
    .limit(limit);
};

// Static method for analytics aggregation
sensorDataSchema.statics.getAggregatedData = function(deviceId, sensorType, period = 'hour') {
  const groupBy = {
    hour: {
      year: { $year: '$createdAt' },
      month: { $month: '$createdAt' },
      day: { $dayOfMonth: '$createdAt' },
      hour: { $hour: '$createdAt' }
    },
    day: {
      year: { $year: '$createdAt' },
      month: { $month: '$createdAt' },
      day: { $dayOfMonth: '$createdAt' }
    },
    week: {
      year: { $year: '$createdAt' },
      week: { $week: '$createdAt' }
    },
    month: {
      year: { $year: '$createdAt' },
      month: { $month: '$createdAt' }
    }
  };

  return this.aggregate([
    {
      $match: {
        deviceId,
        sensorType,
        createdAt: { $gte: new Date(Date.now() - 30 * 24 * 60 * 60 * 1000) }
      }
    },
    {
      $group: {
        _id: groupBy[period],
        avgTemperature: { $avg: '$data.temperature' },
        maxTemperature: { $max: '$data.temperature' },
        minTemperature: { $min: '$data.temperature' },
        avgHumidity: { $avg: '$data.humidity' },
        motionCount: { $sum: { $cond: ['$data.motionDetected', 1, 0] } },
        alertCount: { 
          $sum: { 
            $cond: [{ $in: ['$alertLevel', ['warning', 'critical']] }, 1, 0] 
          } 
        },
        dataPoints: { $sum: 1 },
        timestamp: { $first: '$createdAt' }
      }
    },
    { $sort: { '_id.year': 1, '_id.month': 1, '_id.day': 1, '_id.hour': 1 } }
  ]);
};

module.exports = mongoose.model('SensorData', sensorDataSchema);

File: Lab2/apz-pzpi-22-9-aleksiienko-yevhenii-lab2/data/models/SensorData.kt
package com.smartmonitoringapplication.app.data.models

import android.os.Parcelable
import kotlinx.parcelize.Parcelize

@Parcelize
data class SensorData(
    val _id: String,
    val deviceId: String,
    val userId: String,
    val sensorType: String, // temperature, motion, humidity, system
    val data: SensorReading,
    val alertLevel: String = "normal", // info, warning, critical, normal
    val processed: Boolean = false,
    val processingResult: ProcessingResult? = null,
    val metadata: SensorMetadata? = null,
    val createdAt: String,
    val updatedAt: String
) : Parcelable

@Parcelize
data class SensorReading(
    // Temperature sensor data
    val temperature: Double? = null,
    val humidity: Double? = null,

    // Motion sensor data
    val motionDetected: Boolean? = null,
    val sensorId: Int? = null,
    val location: String? = null,

    // System data
    val battery: Int? = null,
    val signalStrength: Int? = null,
    val memory: MemoryReading? = null,
    val uptime: Long? = null
) : Parcelable

@Parcelize
data class MemoryReading(
    val used: Long,
    val free: Long
) : Parcelable

@Parcelize
data class ProcessingResult(
    val anomaly: Boolean = false,
    val confidence: Double? = null,
    val recommendations: List<String> = emptyList(),
    val trend: String? = null // increasing, decreasing, stable, volatile
) : Parcelable

@Parcelize
data class SensorMetadata(
    val firmware: String? = null,
    val deviceTimestamp: Long? = null,
    val processingTime: Long? = null,
    val dataQuality: String = "good" // excellent, good, poor, invalid
) : Parcelable

// Analytics models
data class AnalyticsResponse(
    val success: Boolean,
    val message: String? = null,
    val data: AnalyticsData? = null
)

data class AnalyticsData(
    val stats: DashboardStats? = null,
    val recentAlerts: List<SensorData>? = null,
    val temperatureTrends: List<TemperatureTrend>? = null,
    val motionActivity: List<MotionActivity>? = null,
    val deviceHealth: List<DeviceHealth>? = null,
    val generatedAt: String? = null,

    // Temperature analytics
    val devices: List<DeviceTemperatureStats>? = null,
    val trends: List<TemperatureTrendDetailed>? = null,
    val statistics: OverallTemperatureStats? = null,
    val anomalies: List<SensorData>? = null,
    val period: AnalyticsPeriod? = null,

    // Motion analytics
    val patterns: List<MotionPattern>? = null,
    val locationActivity: List<LocationActivity>? = null,
    val hourlyDistribution: List<HourlyDistribution>? = null,

    // Alerts analytics
    val summary: AlertsSummary? = null,
    val alertsBySeverity: List<AlertsBySeverity>? = null,
    val alertsByDevice: List<AlertsByDevice>? = null,
    val alertTrends: List<AlertTrend>? = null,
    val recentCriticalAlerts: List<SensorData>? = null,
    val alertResolution: List<AlertResolution>? = null
)

@Parcelize
data class DashboardStats(
    val totalDevices: Int,
    val onlineDevices: Int,
    val dataPointsToday: Int,
    val alertsThisWeek: Int
) : Parcelable

@Parcelize
data class TemperatureTrend(
    val _id: TrendId,
    val avgTemperature: Double,
    val maxTemperature: Double,
    val minTemperature: Double,
    val avgHumidity: Double? = null,
    val dataPoints: Int,
    val alerts: Int
) : Parcelable

@Parcelize
data class TrendId(
    val hour: Int? = null,
    val deviceId: String? = null,
    val period: String? = null
) : Parcelable

@Parcelize
data class MotionActivity(
    val _id: String,
    val motionEvents: Int,
    val totalEvents: Int
) : Parcelable

@Parcelize
data class DeviceHealth(
    val deviceId: String,
    val name: String,
    val location: String,
    val healthScore: Int,
    val isOnline: Boolean,
    val lastSeen: String,
    val battery: Int? = null
) : Parcelable

// Additional analytics models
data class DeviceTemperatureStats(
    val deviceId: String,
    val name: String,
    val location: String,
    val statistics: TemperatureStatistics
)

data class TemperatureStatistics(
    val avgTemperature: Double? = null,
    val maxTemperature: Double? = null,
    val minTemperature: Double? = null,
    val dataPoints: Int = 0,
    val alerts: Int = 0
)

data class TemperatureTrendDetailed(
    val _id: DetailedTrendId,
    val avgTemperature: Double,
    val maxTemperature: Double,
    val minTemperature: Double,
    val avgHumidity: Double? = null,
    val dataPoints: Int,
    val alerts: Int
)

data class DetailedTrendId(
    val deviceId: String,
    val period: String
)

data class OverallTemperatureStats(
    val avgTemperature: Double? = null,
    val maxTemperature: Double? = null,
    val minTemperature: Double? = null,
    val avgHumidity: Double? = null,
    val totalDataPoints: Int = 0,
    val totalAlerts: Int = 0
)

data class AnalyticsPeriod(
    val start: String,
    val end: String,
    val granularity: String
)

// Motion analytics
data class MotionPattern(
    val _id: MotionPatternId,
    val motionEvents: Int,
    val totalEvents: Int
)

data class MotionPatternId(
    val deviceId: String,
    val location: String,
    val period: String
)

data class LocationActivity(
    val _id: String,
    val motionEvents: Int,
    val totalEvents: Int,
    val devices: List<String>
)

data class HourlyDistribution(
    val _id: Int, // hour
    val motionEvents: Int,
    val locations: List<String>
)

// Alerts analytics
data class AlertsSummary(
    val totalAlerts: Int,
    val deviceCount: Int,
    val period: AnalyticsPeriod
)

data class AlertsBySeverity(
    val _id: String, // alertLevel
    val count: Int,
    val devices: List<String>
)

data class AlertsByDevice(
    val _id: String, // deviceId
    val totalAlerts: Int,
    val warningAlerts: Int,
    val criticalAlerts: Int,
    val sensorTypes: List<String>
)

data class AlertTrend(
    val _id: AlertTrendId,
    val count: Int
)

data class AlertTrendId(
    val date: String,
    val alertLevel: String
)

data class AlertResolution(
    val _id: String, // sensorType
    val avgResponseTime: Double? = null,
    val totalAlerts: Int,
    val resolvedAlerts: Int
)

// Response wrapper for sensor data
data class SensorDataResponse(
    val success: Boolean,
    val message: String? = null,
    val data: SensorDataResponseData? = null
)

data class SensorDataResponseData(
    val rawData: List<SensorData>,
    val aggregatedData: List<TemperatureTrend>? = null,
    val period: String? = null,
    val deviceInfo: DeviceInfo? = null
)

data class DeviceInfo(
    val deviceId: String,
    val name: String,
    val location: String
)

// Alerts response
data class AlertsResponse(
    val success: Boolean,
    val message: String? = null,
    val data: AlertsData? = null
)

data class AlertsData(
    val alerts: List<SensorData>,
    val pagination: Pagination
)